### DISCUSSION TITLE
Express composition url and protocol are undefined

### DISCUSSION THREAD

**Discussion by BillyCottrell** [loopbackio/loopback-next#7525] [Q&A]
Hi there,

I can't seem to get the server URL to be displayed, when retrieving the URL of the restServer it is for some reason undefined.
So I thought to use the config parameters to construct the URL myself but again I wasn't able to do it because the protocol was undefined as well for some reason, at first I was thinking that I was doing something wrong but that isn't the case since its literally the express composition repo I am using.
You can get the same result when using the express composition repo and edit the index.ts file main function:
```
export async function main(options: ApplicationConfig = {}) {
  const server = new ExpressServer(options);
  await server.boot();
  await server.start();

  const rsurl = server.lbApp.restServer.url;
  console.log(`Server is running at ${rsurl}`);
  // result: "Server is running at undefined"
  
  const serverConfig = server.lbApp.restServer.config
  const port = serverConfig.port === undefined ? '' : serverConfig.port.toString.length > 0 ? ':' + serverConfig.port : '';
  const url = serverConfig.protocol + '://' + serverConfig.host + port + '/' + serverConfig.basePath;
  console.log(`Server is running at ${url}`);
  // result: "Server is running at undefined://localhost:3000/"
}
```

Am I doing something wrong or is it just not possible because of the express server?

Thanks in advance,

Billy Cottrell

**Comment by achrinza** [comment]
From my understanding, `restServer` class member is provided by `RestApplication` from `@loopback/rest`.

For the `ExpressServer` in `@loopback/express`, this is provided via `url` variable:

```typescript
cont server = new ExpressServer();
const url = server.url;
```

**Reply by BillyCottrell** [reply]
Express Server is a custom class made in server.ts and has the following code:
```
import {once} from 'events';
import express, {Request, Response} from 'express';
import http from 'http';
import {ApplicationConfig, RoomBookingApiApplication} from './application';

export class ExpressServer {
    public readonly app: express.Application;
    public readonly lbApp: RoomBookingApiApplication;
    private server?: http.Server;

    constructor(options: ApplicationConfig = {}) {
        this.app = express()
        this.lbApp = new RoomBookingApiApplication(options);
        this.app.use('/api', this.lbApp.requestHandler);

        this.app.get('/hello', function (_req: Request, res: Response) {
            res.send('Hello world!');
        });
    }

    public async boot() {
        await this.lbApp.boot();
    }

    public async start() {
        await this.lbApp.start();
        const port = this.lbApp.restServer.config.port ?? 3000;
        const host = this.lbApp.restServer.config.host ?? '127.0.0.1';
        this.server = this.app.listen(port, host);
        await once(this.server, 'listening');
    }

    // For testing purposes
    public async stop() {
        if (!this.server) return;
        await this.lbApp.stop();
        this.server.close();
        await once(this.server, 'close');
        this.server = undefined;
    }
}
```

So as you can see if I would use the code you mentioned it wouldn't work since url doesn't exist. Even doing server.app.url won't work, hence the reason why I posted this question, since I don't know where I am able to retrieve this information.

The reason I have set it up like this is because I am trying to develop middleware purely to be able to communicate with Microsoft Graph and retrieve booking information from a resource mailbox. Also, because I am used to building lb3 apps but not lb4 I thought it would be a good idea to manually convert my lb3 app (which is unfinished) to lb4 and complete the app. Except that I don't know how to convert my old middleware to the new lb4. So I thought to start by building an express server inside the lb4 app and work my way through it to implement the whole code base (it's not that much, luckily).

The second reason is that I am trying to convert it into a microservice that can auto register itself at the API gateway so I really need to be able to retrieve the protocol, host and port, and it can't be the value undefined.

**Reply by BillyCottrell** [reply]
any ideas on this?

Thanks in advance!
